File: Cracking the coding interviews/Linked Lists/LinkedListOperation/LinkedList.java
class Node {
    int data;
    Node next;

    public Node(int value) {
        this.data = value;
        this.next = null;
    }
}

public class LinkedList {
    Node head;

    // Append value to the end of the list iteratively
    public void appendToTail(int value) {
        if (head == null) {
            head = new Node(value);
            // Note: this return is important as it stops adding one extra node into the
            // linked list
            return;
        }
        Node currNode = head;
        while (currNode.next != null) {
            currNode = currNode.next;
        }
        currNode.next = new Node(value);
    }

    public void prepend(int value) {
        Node newNode = new Node(value);
        newNode.next = head;
        head = newNode;
    }

    public void delete(int value) {
        if (head == null)
            return;

        if (head.data == value) {
            head = head.next;
        }

        Node currNode = head;
        while (currNode.next != null) {
            if (currNode.next.data == value) {
                // This is going to skip the first value appearance
                currNode.next = currNode.next.next;
                // Note: this return here stops the search.
                // Include return if only want to delete the first appearance of the node.
                return;
            }
            currNode = currNode.next;
        }
    }

    // Print the linked list iteratively
    public void printLinkedList() {
        Node n = head;
        while (n != null) {
            System.out.print(n.data + " ");
            n = n.next;
        }
        System.out.println();
    }

    // Append value to the end of the list recursively
    public void appendToTailRecursion(Node node, int value) {
        if (node == null) {
            head = new Node(value); // If head is null, initialize the list
        } else if (node.next == null) {
            node.next = new Node(value);
        } else {
            appendToTailRecursion(node.next, value);
        }
    }

    // Print the linked list recursively
    public void printLinkedListRecursion(Node node) {
        if (node == null) {
            return;
        }
        System.out.print(node.data + " ");
        printLinkedListRecursion(node.next);
    }
}

File: Cracking the coding interviews/Linked Lists/LinkedListOperationCSharp/utils/LinkedList.cs
namespace LinkedListOperationCSharp.LinkedList
{
    public class LinkedListNode
    {
        public int data { get; set; }
        public LinkedListNode? next { get; set; }

        public LinkedListNode(int value)
        {
            data = value;
            next = null;
        }
    }

    public class LinkedList
    {
        public LinkedListNode? head;

        public void appendToTail(int value)
        {
            if (head == null)
            {
                head = new LinkedListNode(value);
                return;
            }

            LinkedListNode currNode = head;

            while (currNode.next != null)
            {
                currNode = currNode.next;
            }

            currNode.next = new LinkedListNode(value);
        }

        public void prepend(int value)
        {
            LinkedListNode newNode = new LinkedListNode(value);
            newNode.next = head;
            head = newNode;
        }

        public void delete(int value)
        {
            if (head == null) return;

            if (head.data == value)
            {
                head = head.next;
                return;
            }

            LinkedListNode currNode = head;
            while (currNode.next != null)
            {
                if (currNode.next.data == value)
                {
                    currNode.next = currNode.next.next;
                    return;
                }

                currNode = currNode.next;
            }
        }

        public bool hasCycle()
        {
            if (head == null || head.next == null) return false;

            LinkedListNode? slow = head;
            LinkedListNode? fast = head;

            while (slow != null && fast != null && fast.next != null)
            {
                slow = slow.next;
                fast = fast.next.next;

                if (slow == fast)
                {
                    return true;
                }
            }
            return false;
        }

        public void makeCycle()
        {
            if (head == null) return;

            LinkedListNode currNode = head;
            while (currNode.next != null)
            {
                currNode = currNode.next;
            }

            currNode.next = head;
        }

        public void printLinkedList()
        {
            if (head == null) return;

            LinkedListNode? currNode = head;
            while (currNode != null)
            {
                Console.Write(currNode.data + " ");
                currNode = currNode.next;
            }
        }

        public void appendToTailRecursion(LinkedListNode? node, int value)
        {
            if (node == null)
            {
                head = new LinkedListNode(value);
                return;
            }

            if (node.next == null)
            {
                node.next = new LinkedListNode(value);
                return;
            }

            appendToTailRecursion(node.next, value);
        }

        public void printLinkedListRecursion(LinkedListNode? node)
        {
            if (node == null)
            {
                return;
            }

            Console.Write(node.data + " ");
            printLinkedListRecursion(node.next);
        }
    }
}